src/main/java/org/springframework/data/jpa/repository/query/JSqlParserQueryUtils.java

```diff
@@ -0,0 +1,462 @@
+	 */
+	public static String getExistsQueryString(String entityName, String countQueryPlaceHolder,
+			Iterable<String> idAttributes) {
+
+		final Table tableNameWithAlias = getTableWithAlias(entityName, DEFAULT_TABLE_ALIAS);
+		Function jSqlCount = getJSqlCount(Collections.singletonList(countQueryPlaceHolder), false);
+
+		Select select = SelectUtils.buildSelectFromTableAndSelectItems(tableNameWithAlias,
+				new SelectExpressionItem(jSqlCount));
+
+		PlainSelect selectBody = (PlainSelect) select.getSelectBody();
+
+		List<Expression> equalityExpressions = Streamable.of(idAttributes).stream() //
+				.map(field -> {
+					Expression tableNameField = new Column().withTable(tableNameWithAlias).withColumnName(field);
+					Expression inputField = new Column(":".concat(field));
+					return new EqualsTo(tableNameField, inputField);
+				}).collect(Collectors.toList());
+
+		if (equalityExpressions.size() > 1) {
+			AndExpression rootOfWhereClause = concatenateWithAndExpression(equalityExpressions);
+			selectBody.setWhere(rootOfWhereClause);
+		} else if (equalityExpressions.size() == 1) {
+			selectBody.setWhere(equalityExpressions.get(0));
+		}
+
+		return selectBody.toString();
+	}
+
+	/**
+	 * Returns the query string for the given class name.
+	 *
+	 * @param template must not be {@literal null}.
+	 * @param entityName must not be {@literal null}.
+	 * @return the template with placeholders replaced by the {@literal entityName}. Guaranteed to be not {@literal null}.
+	 */
+	public static String getQueryString(String template, String entityName) {
+		Assert.hasText(entityName, "Entity name must not be null or empty!");
+		return String.format(template, entityName);
+	}
+
+	/**
+	 * Resolves the alias for the entity to be retrieved from the given JPA query. Note that you only provide valid Query
+	 * strings. Things such as <code>from User u</code> will throw an {@link IllegalArgumentException}.
+	 * 
+	 * @param query must not be {@literal null}.
+	 * @return Might return {@literal null}.
+	 */
+	public static String detectAlias(String query) {
+		Select selectStatement = parseSelectStatement(query);
+		PlainSelect selectBody = (PlainSelect) selectStatement.getSelectBody();
+		return detectAlias(selectBody);
+	}
+
+	/**
+	 * Resolves the alias for the entity to be retrieved from the given {@link PlainSelect}. Note that you only provide
+	 * valid Query strings. Things such as <code>from User u</code> will throw an {@link IllegalArgumentException}.
+	 *
+	 * @param selectBody must not be {@literal null}.
+	 * @return Might return {@literal null}.
+	 */
+	private static String detectAlias(PlainSelect selectBody) {
+		Alias alias = selectBody.getFromItem().getAlias();
+		return alias == null ? null : alias.getName();
+	}
+
+	/**
+	 * Parses a query string with JSqlParser.
+	 * 
+	 * @param query the query to parse
+	 * @return the parsed query
+	 */
+	private static Select parseSelectStatement(String query) {
```

Comment:
During debugging I realized this and already tried to memorize inside `StringQuery`. Although the problem is that JSqlParser currently does not support copy by value with a method and therefor if I change the selectbody (for example for adding a "order by" clause ) the "memorized query" will be changed too.

I could add my own "copy by value" method but this may be not future prove.

Reply:
I raised an issue at JSqlParser maybe the have any idea to "bypass" the copy by reference. 
https://github.com/JSQLParser/JSqlParser/issues/1468

Reply:
Ok, then we'll leave it as it is for now.

Reply:
In my opinion the best idea would be to create a whole issue for this "optimization" and revisit it as soon as a better solution is provided directly by JSqlparser or we somehow can "bypass" this situation.